pkg/discovery/dtofactory/cluster_dto_builder.go

```diff
@@ -51,7 +52,81 @@ func (builder *clusterDTOBuilder) Build() []*proto.GroupDTO {
 	if err != nil {
 		glog.Errorf("Failed to build VM group DTO %v", err)
 	} else {
-		clusterDTOs = append(clusterDTOs, vmGroupDTO)
+		clusterGroupDTOs = append(clusterGroupDTOs, vmGroupDTO)
 	}
-	return clusterDTOs
+	return clusterGroupDTOs
+}
+
+func (builder *clusterDTOBuilder) BuildEntity(entityDTOs []*proto.EntityDTO) (*proto.EntityDTO, error) {
+	// id.
+	uid := builder.cluster.Name
+	entityDTOBuilder := sdkbuilder.NewEntityDTOBuilder(proto.EntityDTO_K8S_CLUSTER, uid)
+
+	// display name.
+	displayName := builder.targetId
+	entityDTOBuilder.DisplayName(displayName)
+
+	// Resource commodities sold.
+	commoditiesSold, err := builder.getCommoditiesSold(entityDTOs)
+	if err != nil {
+		glog.Errorf("Error creating commoditiesSold for %s: %s", builder.cluster.Name, err)
+		return nil, err
+	}
+	entityDTOBuilder.SellsCommodities(commoditiesSold)
+
+	// Cluster entity cannot be provisioned or suspended by Turbonomic analysis
+	entityDTOBuilder.IsProvisionable(false)
+	entityDTOBuilder.IsSuspendable(false)
+
+	entityDTOBuilder.WithPowerState(proto.EntityDTO_POWERED_ON)
+
+	// Link member nodes as connected entities
+	for _, node := range builder.cluster.NodeList {
+		entityDTOBuilder.Owns(string(node.UID))
+	}
+
+	// build entityDTO.
+	entityDto, err := entityDTOBuilder.Create()
+	if err != nil {
+		glog.Errorf("Failed to build Cluster entityDTO: %s", err)
+		return nil, err
+	}
+
+	glog.V(4).Infof("Cluster DTO: %+v", entityDto)
+	return entityDto, nil
+}
+
+func (builder *clusterDTOBuilder) getCommoditiesSold(entityDTOs []*proto.EntityDTO) ([]*proto.CommodityDTO, error) {
+	// Accumulate used and capacity values from the nodes on the cluster
+	used := make(map[proto.CommodityDTO_CommodityType]float64)
+	capacity := make(map[proto.CommodityDTO_CommodityType]float64)
+	for _, entityDTO := range entityDTOs {
+		if *entityDTO.EntityType == proto.EntityDTO_VIRTUAL_MACHINE {
+			for _, commodity := range entityDTO.GetCommoditiesSold() {
+				if commodity.Used != nil && commodity.Capacity != nil {
+					used[*commodity.CommodityType] = used[*commodity.CommodityType] + *commodity.Used
```

Comment:
Yes, you read it right.  Definitely arguable, but I chose to aggregate everything.  Some may not be of any use now, but on one hand there's just one cluster per probe some a few more commodities for the cluster won't cause performance issues.  On the other hand, even aggregating the access commodity may provide some value, e.g. how many nodes in the cluster with no schedule set, etc.

If we want to filter out all the access commodities, then we will have to explicitly manage yet another list (of access commodities), because `CommodityType` has no method called `isAccessCommodity`.  I feel that's more code to manage unnecessarily.  But I'm open to more opinions.

Reply:
I think its fine to have it as it is now. As you rightly mentioned its not really a performance overhead.